Bedrock.Framework/Protocols/Protocol.cs

```diff
@@ -53,40 +66,25 @@ public async ValueTask<TReadMessage> ReadAsync(CancellationToken cancellationTok
                         // No message limit, just parse and dispatch
                         if (_maximumMessageSize == null)
                         {
-                            if (reader.TryParseMessage(buffer, out consumed, out examined, out protocolMessage))
+                            while (reader.TryParseMessage(buffer, out consumed, out examined, out var protocolMessage))
                             {
-                                return protocolMessage;
+                                await inputBuffer.WriteAsync(protocolMessage);
+                                buffer = buffer.Slice(consumed);
                             }
                         }
                         else
                         {
                             // We give the parser a sliding window of the default message size
                             var maxMessageSize = _maximumMessageSize.Value;
+                            if (buffer.Length > maxMessageSize)
+                            {
+                                inputBuffer.Complete(new InvalidDataException($"The maximum message size of {maxMessageSize}B was exceeded. The message size can be configured in AddHubOptions."));
+                            }
 
-                            if (!buffer.IsEmpty)
+                            while (reader.TryParseMessage(buffer, out consumed, out examined, out var protocolMessage))
                             {
-                                var segment = buffer;
```

Comment:
Why did you remove this logic?

Reply:
the same check is in line 62 so i think its redundant

Reply:
regarding the other comment: I reverted the change in the branch with msg size limit seems I overlooked something while trying to simplify things